Still getting a ```--- FAIL: TestWork (0.00s)
    --- FAIL: TestWork/kube_worker/Reload_mesh_ (39.46s)
        work_test.go:664: write unix @->/tmp/receptor-testing/controlsockets/091534067/controlsock: write: broken pipe```

tried to do a ```close()``` and ```reconnect()``` but that causes another error.

--- a/Makefile
+++ b/Makefile
@@ -78,7 +78,7 @@ TESTCMD = -run $(RUNTEST)
 endif
 
 test:
-	@go test ./... -p 1 -parallel=16 $(TESTCMD) -count=1
+	@go test ./... -p 1 -parallel=16 $(TESTCMD) -count=1 -v
 
 testloop: receptor
 	@i=1; while echo "------ $$i" && \

--- a/tests/functional/lib/mesh/climesh.go
+++ b/tests/functional/lib/mesh/climesh.go
@@ -150,6 +150,111 @@ func (m *CLIMesh) Nodes() map[string]Node {
 	return nodes
 }
 
+func createNodedefConnections(meshDefinition *YamlData, existingMesh *CLIMesh) error {
+	for k := range meshDefinition.Nodes {
+		for connNode, connYaml := range meshDefinition.Nodes[k].Connections {
+			index := connYaml.Index
+			TLS := connYaml.TLS
+			var attr interface{}
+			if existingMesh != nil {
+				attr = existingMesh.MeshDefinition.Nodes[connNode].NodedefBase[index]
+			} else {
+				attr = meshDefinition.Nodes[connNode].NodedefBase[index]
+			}
+			attrMap, _ := attr.(map[interface{}]interface{})
+			listener, ok := attrMap["tcp-listener"]
+			if ok {
+				dialerYaml := make(map[interface{}]interface{})
+				listenerMap, ok := listener.(map[interface{}]interface{})
+				if !ok {
+					return errors.New("listener object is not a map")
+				}
+				peerYaml := make(map[interface{}]interface{})
+				bindaddr, ok := listenerMap["bindaddr"].(string)
+				var addr string
+				if ok {
+					addr = bindaddr + ":" + listenerMap["port"].(string)
+				} else {
+					addr = "localhost:" + listenerMap["port"].(string)
+				}
+				peerYaml["address"] = addr
+				peerYaml["cost"] = getListenerCost(listenerMap, k)
+
+				if TLS != "" {
+					peerYaml["tls"] = TLS
+				}
+				dialerYaml["tcp-peer"] = peerYaml
+				if existingMesh != nil {
+					existingMesh.MeshDefinition.Nodes[k].NodedefConnections = append(existingMesh.MeshDefinition.Nodes[k].NodedefConnections, dialerYaml)
+				} else {
+					meshDefinition.Nodes[k].NodedefConnections = append(meshDefinition.Nodes[k].NodedefConnections, dialerYaml)
+				}
+			}
+			listener, ok = attrMap["udp-listener"]
+			if ok {
+				dialerYaml := make(map[interface{}]interface{})
+				listenerMap, ok := listener.(map[interface{}]interface{})
+				if !ok {
+					return errors.New("listener object is not a map")
+				}
+				peerYaml := make(map[interface{}]interface{})
+				bindaddr, ok := listenerMap["bindaddr"].(string)
+				var addr string
+				if ok {
+					addr = bindaddr + ":" + listenerMap["port"].(string)
+				} else {
+					addr = "localhost:" + listenerMap["port"].(string)
+				}
+				peerYaml["address"] = addr
+				peerYaml["cost"] = getListenerCost(listenerMap, k)
+				dialerYaml["udp-peer"] = peerYaml
+				if existingMesh != nil {
+					existingMesh.MeshDefinition.Nodes[k].NodedefConnections = append(existingMesh.MeshDefinition.Nodes[k].NodedefConnections, dialerYaml)
+				} else {
+					meshDefinition.Nodes[k].NodedefConnections = append(meshDefinition.Nodes[k].NodedefConnections, dialerYaml)
+				}
+			}
+
+			listener, ok = attrMap["ws-listener"]
+			if ok {
+				dialerYaml := make(map[interface{}]interface{})
+				listenerMap, ok := listener.(map[interface{}]interface{})
+				if !ok {
+					return errors.New("listener object is not a map")
+				}
+				peerYaml := make(map[interface{}]interface{})
+
+				proto := "ws://"
+				tlsName, tlsEnabled := listenerMap["tls"].(string)
+				if tlsEnabled && tlsName != "" {
+					proto = "wss://"
+				}
+
+				bindaddr, ok := listenerMap["bindaddr"].(string)
+				var addr string
+				if ok {
+					addr = proto + bindaddr + ":" + listenerMap["port"].(string)
+				} else {
+					addr = proto + "localhost:" + listenerMap["port"].(string)
+				}
+				peerYaml["address"] = addr
+				peerYaml["cost"] = getListenerCost(listenerMap, k)
+				if TLS != "" {
+					peerYaml["tls"] = TLS
+				}
+				dialerYaml["ws-peer"] = peerYaml
+				if existingMesh != nil {
+					existingMesh.MeshDefinition.Nodes[k].NodedefConnections = append(existingMesh.MeshDefinition.Nodes[k].NodedefConnections, dialerYaml)
+				} else {
+					meshDefinition.Nodes[k].NodedefConnections = append(meshDefinition.Nodes[k].NodedefConnections, dialerYaml)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
 // NewCLIMeshFromFile Takes a filename of a file with a yaml description of a mesh, loads it and
 // calls NewMeshFromYaml on it.
 func NewCLIMeshFromFile(filename, dirSuffix string) (Mesh, error) {
@@ -202,7 +307,7 @@ func NewCLIMeshFromYaml(meshDefinition YamlData, dirSuffix string) (*CLIMesh, er
 		// Keep track of if we need to add an attribute for the node id or if
 		// it already exists
 		needsIDAttr := true
-		for attrkey, attr := range meshDefinition.Nodes[k].Nodedef {
+		for attrkey, attr := range meshDefinition.Nodes[k].NodedefBase {
 			attrMap := attr.(map[interface{}]interface{})
 			for k, v := range attrMap {
 				k = k.(string)
@@ -232,7 +337,7 @@ func NewCLIMeshFromYaml(meshDefinition YamlData, dirSuffix string) (*CLIMesh, er
 					}
 				}
 			}
-			meshDefinition.Nodes[k].Nodedef[attrkey] = attrMap
+			meshDefinition.Nodes[k].NodedefBase[attrkey] = attrMap
 		}
 		if needsIDAttr {
 			idYaml := make(map[interface{}]interface{})
@@ -241,103 +346,26 @@ func NewCLIMeshFromYaml(meshDefinition YamlData, dirSuffix string) (*CLIMesh, er
 			nodeYaml["datadir"] = filepath.Join(node.dir, "datadir")
 			os.Mkdir(nodeYaml["datadir"].(string), 0o755)
 			idYaml["node"] = nodeYaml
-			meshDefinition.Nodes[k].Nodedef = append(meshDefinition.Nodes[k].Nodedef, idYaml)
+			meshDefinition.Nodes[k].NodedefBase = append(meshDefinition.Nodes[k].NodedefBase, idYaml)
 		}
 		logYaml := make(map[interface{}]interface{})
 		levelYaml := make(map[interface{}]interface{})
 		levelYaml["level"] = "debug"
 		logYaml["log-level"] = levelYaml
-		meshDefinition.Nodes[k].Nodedef = append(meshDefinition.Nodes[k].Nodedef, logYaml)
+		meshDefinition.Nodes[k].NodedefBase = append(meshDefinition.Nodes[k].NodedefBase, logYaml)
 		nodes[k] = node
 	}
-	for k := range meshDefinition.Nodes {
-		for connNode, connYaml := range meshDefinition.Nodes[k].Connections {
-			index := connYaml.Index
-			TLS := connYaml.TLS
-			attr := meshDefinition.Nodes[connNode].Nodedef[index]
-			attrMap := attr.(map[interface{}]interface{})
-			listener, ok := attrMap["tcp-listener"]
-			if ok {
-				dialerYaml := make(map[interface{}]interface{})
-				listenerMap, ok := listener.(map[interface{}]interface{})
-				if !ok {
-					return nil, errors.New("listener object is not a map")
-				}
-				peerYaml := make(map[interface{}]interface{})
-				bindaddr, ok := listenerMap["bindaddr"].(string)
-				var addr string
-				if ok {
-					addr = bindaddr + ":" + listenerMap["port"].(string)
-				} else {
-					addr = "localhost:" + listenerMap["port"].(string)
-				}
-				peerYaml["address"] = addr
-				peerYaml["cost"] = getListenerCost(listenerMap, k)
 
-				if TLS != "" {
-					peerYaml["tls"] = TLS
-				}
-				dialerYaml["tcp-peer"] = peerYaml
-				meshDefinition.Nodes[k].Nodedef = append(meshDefinition.Nodes[k].Nodedef, dialerYaml)
-			}
-			listener, ok = attrMap["udp-listener"]
-			if ok {
-				dialerYaml := make(map[interface{}]interface{})
-				listenerMap, ok := listener.(map[interface{}]interface{})
-				if !ok {
-					return nil, errors.New("listener object is not a map")
-				}
-				peerYaml := make(map[interface{}]interface{})
-				bindaddr, ok := listenerMap["bindaddr"].(string)
-				var addr string
-				if ok {
-					addr = bindaddr + ":" + listenerMap["port"].(string)
-				} else {
-					addr = "localhost:" + listenerMap["port"].(string)
-				}
-				peerYaml["address"] = addr
-				peerYaml["cost"] = getListenerCost(listenerMap, k)
-				dialerYaml["udp-peer"] = peerYaml
-				meshDefinition.Nodes[k].Nodedef = append(meshDefinition.Nodes[k].Nodedef, dialerYaml)
-			}
-			listener, ok = attrMap["ws-listener"]
-			if ok {
-				dialerYaml := make(map[interface{}]interface{})
-				listenerMap, ok := listener.(map[interface{}]interface{})
-				if !ok {
-					return nil, errors.New("listener object is not a map")
-				}
-				peerYaml := make(map[interface{}]interface{})
-
-				proto := "ws://"
-				tlsName, tlsEnabled := listenerMap["tls"].(string)
-				if tlsEnabled && tlsName != "" {
-					proto = "wss://"
-				}
-
-				bindaddr, ok := listenerMap["bindaddr"].(string)
-				var addr string
-				if ok {
-					addr = proto + bindaddr + ":" + listenerMap["port"].(string)
-				} else {
-					addr = proto + "localhost:" + listenerMap["port"].(string)
-				}
-				peerYaml["address"] = addr
-				peerYaml["cost"] = getListenerCost(listenerMap, k)
-				if TLS != "" {
-					peerYaml["tls"] = TLS
-				}
-				dialerYaml["ws-peer"] = peerYaml
-				meshDefinition.Nodes[k].Nodedef = append(meshDefinition.Nodes[k].Nodedef, dialerYaml)
-			}
-		}
+	err = createNodedefConnections(&meshDefinition, nil)
+	if err != nil {
+		return nil, err
 	}
 
 	// Setup the controlsvc and sockets
 	for k, node := range nodes {
 		needsControlService := true
 		controlServiceIndex := 0
-		for index, attr := range meshDefinition.Nodes[k].Nodedef {
+		for index, attr := range meshDefinition.Nodes[k].NodedefBase {
 			attrMap := attr.(map[interface{}]interface{})
 			for k, v := range attrMap {
 				k = k.(string)
@@ -368,14 +396,26 @@ func NewCLIMeshFromYaml(meshDefinition YamlData, dirSuffix string) (*CLIMesh, er
 			tmp["filename"] = controlSocket
 			controlServiceYaml := make(map[interface{}]interface{})
 			controlServiceYaml["control-service"] = tmp
-			meshDefinition.Nodes[k].Nodedef = append(meshDefinition.Nodes[k].Nodedef, controlServiceYaml)
+			meshDefinition.Nodes[k].NodedefBase = append(meshDefinition.Nodes[k].NodedefBase, controlServiceYaml)
 		} else {
-			meshDefinition.Nodes[k].Nodedef[controlServiceIndex].(map[interface{}]interface{})["control-service"].(map[interface{}]interface{})["filename"] = controlSocket
+			meshDefinition.Nodes[k].NodedefBase[controlServiceIndex].(map[interface{}]interface{})["control-service"].(map[interface{}]interface{})["filename"] = controlSocket
 		}
 	}
 
+	// nodedef = NodedefBase
+	for k := range meshDefinition.Nodes {
+		if meshDefinition.Nodes[k].NodedefBase != nil {
+			// trying to avoid = due to pointer issues in []interface{}
+			meshDefinition.Nodes[k].Nodedef = append(meshDefinition.Nodes[k].Nodedef, meshDefinition.Nodes[k].NodedefBase...)
+		}
+	}
+
+	// yamlConfig = Nodedef+NodedefConnections
 	for k, node := range nodes {
 		node.yamlConfig = meshDefinition.Nodes[k].Nodedef
+		if meshDefinition.Nodes[k].NodedefConnections != nil {
+			node.yamlConfig = append(node.yamlConfig, meshDefinition.Nodes[k].NodedefConnections...)
+		}
 		err = node.Start()
 		if err != nil {
 			return nil, err
@@ -406,6 +446,54 @@ func NewCLIMeshFromYaml(meshDefinition YamlData, dirSuffix string) (*CLIMesh, er
 	}
 
 	return mesh, nil
+}
+
+func ModifyCLIMeshFromYaml(meshDefinition YamlData, existingMesh CLIMesh) error {
+	// setting the current `yamlConfig` for each node as nil
+	for existingNodes := range existingMesh.nodes {
+		existingMesh.nodes[existingNodes].yamlConfig = nil
+	}
+
+	// clear out `[]Nodedef{}`
+	for nodes := range existingMesh.MeshDefinition.Nodes {
+		existingMesh.MeshDefinition.Nodes[nodes].Nodedef = nil
+	}
+
+	// setting `[]NodedefConnections{}` to nil in existing mesh
+	for nodes := range existingMesh.MeshDefinition.Nodes {
+		existingMesh.MeshDefinition.Nodes[nodes].NodedefConnections = nil
+	}
+
+	err := createNodedefConnections(&meshDefinition, &existingMesh)
+	if err != nil {
+		return err
+	}
+
+	// Now we combine NodedefConnections` and `NodedefBase` and assign it to Nodedef
+	for k := range existingMesh.MeshDefinition.Nodes {
+		if existingMesh.MeshDefinition.Nodes[k].NodedefBase != nil {
+			existingMesh.MeshDefinition.Nodes[k].Nodedef = append(existingMesh.MeshDefinition.Nodes[k].Nodedef, existingMesh.MeshDefinition.Nodes[k].NodedefBase...)
+		}
+		if existingMesh.MeshDefinition.Nodes[k].NodedefConnections != nil {
+			existingMesh.MeshDefinition.Nodes[k].Nodedef = append(existingMesh.MeshDefinition.Nodes[k].Nodedef, existingMesh.MeshDefinition.Nodes[k].NodedefConnections...)
+		}
+	}
+
+	// Now we combine `[]NodedefConnections{}` and `[]Nodedef{}` for each node in the mesh and write to disk
+	for k, node := range existingMesh.nodes {
+		node.yamlConfig = existingMesh.MeshDefinition.Nodes[k].Nodedef
+		strData, err := yaml.Marshal(node.yamlConfig)
+		if err != nil {
+			return err
+		}
+		nodedefPath := filepath.Join(node.dir, "nodedef.yaml")
+		err = ioutil.WriteFile(nodedefPath, strData, 0o644)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
 }
 
 // Destroy stops all running Netceptors and their backends and frees all

--- a/tests/functional/lib/mesh/libmesh.go
+++ b/tests/functional/lib/mesh/libmesh.go
@@ -230,7 +230,7 @@ func NewLibMeshFromYaml(meshDefinition YamlData, dirSuffix string) (*LibMesh, er
 			return nil, err
 		}
 		os.Mkdir(node.dir, 0o755)
-		for _, attr := range meshDefinition.Nodes[k].Nodedef {
+		for _, attr := range meshDefinition.Nodes[k].NodedefBase {
 			attrMap := attr.(map[interface{}]interface{})
 			for k, v := range attrMap {
 				k = k.(string)
@@ -321,7 +321,7 @@ func NewLibMeshFromYaml(meshDefinition YamlData, dirSuffix string) (*LibMesh, er
 			}
 		}
 
-		for attrkey, attr := range meshDefinition.Nodes[k].Nodedef {
+		for attrkey, attr := range meshDefinition.Nodes[k].NodedefBase {
 			attrMap := attr.(map[interface{}]interface{})
 			for k, v := range attrMap {
 				k = k.(string)
@@ -411,7 +411,7 @@ func NewLibMeshFromYaml(meshDefinition YamlData, dirSuffix string) (*LibMesh, er
 					}
 				}
 			}
-			meshDefinition.Nodes[k].Nodedef[attrkey] = attrMap
+			meshDefinition.Nodes[k].NodedefBase[attrkey] = attrMap
 		}
 		nodes[k] = node
 	}
@@ -419,7 +419,7 @@ func NewLibMeshFromYaml(meshDefinition YamlData, dirSuffix string) (*LibMesh, er
 		node := nodes[k]
 		for connNode, connYaml := range meshDefinition.Nodes[k].Connections {
 			index := connYaml.Index
-			attr := meshDefinition.Nodes[connNode].Nodedef[index]
+			attr := meshDefinition.Nodes[connNode].NodedefBase[index]
 			attrMap := attr.(map[interface{}]interface{})
 			listener, ok := attrMap["tcp-listener"]
 			if ok {
@@ -537,7 +537,7 @@ func (m *LibMesh) CheckConnections() bool {
 		expectedConnections := map[string]float64{}
 		for k, connYaml := range m.MeshDefinition.Nodes[status.NodeID].Connections {
 			index := connYaml.Index
-			configItemYaml := m.MeshDefinition.Nodes[k].Nodedef[index].(map[interface{}]interface{})
+			configItemYaml := m.MeshDefinition.Nodes[k].NodedefBase[index].(map[interface{}]interface{})
 			listenerYaml, ok := configItemYaml["tcp-listener"].(map[interface{}]interface{})
 			if ok {
 				expectedConnections[k] = getListenerCost(listenerYaml, status.NodeID)

--- a/tests/functional/lib/mesh/mesh.go
+++ b/tests/functional/lib/mesh/mesh.go
@@ -64,9 +64,11 @@ type TCRuleYaml struct {
 
 // YamlNode describes how a single node should be represented in yaml.
 type YamlNode struct {
-	Connections map[string]YamlConnection
-	TCRules     *TCRuleYaml
-	Nodedef     []interface{}
+	Connections        map[string]YamlConnection
+	TCRules            *TCRuleYaml
+	Nodedef            []interface{}
+	NodedefConnections []interface{}
+	NodedefBase        []interface{}
 }
 
 func getListenerCost(listenerYaml map[interface{}]interface{}, nodeID string) float64 {

--- a/tests/functional/lib/receptorcontrol/receptorcontrol.go
+++ b/tests/functional/lib/receptorcontrol/receptorcontrol.go
@@ -52,6 +52,22 @@ func (r *ReceptorControl) Connect(filename string) error {
 	return nil
 }
 
+func (r *ReceptorControl) Reload() error {
+	if _, err := r.WriteStr("reload \n"); err != nil {
+		return err
+	}
+	jsonData, err := r.ReadAndParseJSON()
+	if err != nil {
+		return err
+	}
+	success := jsonData["Success"].(bool)
+	if !success {
+		return errors.New("Error")
+	}
+
+	return nil
+}
+
 // Reconnect to unix socket.
 func (r *ReceptorControl) Reconnect() error {
 	if r.socketFilename != "" {
@@ -317,6 +333,25 @@ func (r *ReceptorControl) assertWorkState(ctx context.Context, unitID string, st
 	}
 
 	return assertWithTimeout(ctx, check)
+}
+
+func (r *ReceptorControl) assertWorkSize(ctx context.Context, unitID string, size int64) bool {
+	check := func() bool {
+		workStatus, _ := r.GetWorkStatus(unitID)
+
+		return workStatus.StdoutSize > size
+	}
+
+	return assertWithTimeout(ctx, check)
+}
+
+// assert if the work size is increasing.
+func (r *ReceptorControl) AssertWorkSizeIncreasing(ctx context.Context, unitID string, size int64) error {
+	if !r.assertWorkSize(ctx, unitID, size) {
+		return fmt.Errorf("failed to assert %s is increasing or ctx timed out", unitID)
+	}
+
+	return nil
 }
 
 // AssertWorkRunning waits until work status is running.

--- a/tests/functional/mesh/mesh-definitions/flat-mesh-tcp.yaml
+++ b/tests/functional/mesh/mesh-definitions/flat-mesh-tcp.yaml
@@ -2,77 +2,77 @@
 nodes:
   controller:
     connections: {}
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node1:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node10:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node11:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node12:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node2:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node3:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node4:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node5:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node6:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node7:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node8:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node9:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefBase:
       - tcp-listener: {}

--- a/tests/functional/mesh/mesh-definitions/flat-mesh-udp.yaml
+++ b/tests/functional/mesh/mesh-definitions/flat-mesh-udp.yaml
@@ -2,77 +2,77 @@
 nodes:
   controller:
     connections: {}
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node1:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node10:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node11:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node12:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node2:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node3:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node4:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node5:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node6:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node7:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node8:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node9:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}

--- a/tests/functional/mesh/mesh-definitions/flat-mesh-ws.yaml
+++ b/tests/functional/mesh/mesh-definitions/flat-mesh-ws.yaml
@@ -2,77 +2,77 @@
 nodes:
   controller:
     connections: {}
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node1:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node10:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node11:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node12:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node2:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node3:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node4:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node5:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node6:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node7:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node8:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node9:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}

--- a/tests/functional/mesh/mesh-definitions/random-mesh-tcp.yaml
+++ b/tests/functional/mesh/mesh-definitions/random-mesh-tcp.yaml
@@ -2,13 +2,13 @@
 nodes:
   controller:
     connections: {}
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node1:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node10:
     connections:
@@ -16,13 +16,13 @@ nodes:
         index: 0
       node4:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node11:
     connections:
       node1:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node12:
     connections:
@@ -30,13 +30,13 @@ nodes:
         index: 0
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node2:
     connections:
       node1:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node3:
     connections:
@@ -44,7 +44,7 @@ nodes:
         index: 0
       node4:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node4:
     connections:
@@ -52,7 +52,7 @@ nodes:
         index: 0
       node2:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node5:
     connections:
@@ -60,7 +60,7 @@ nodes:
         index: 0
       node12:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node6:
     connections:
@@ -68,7 +68,7 @@ nodes:
         index: 0
       node3:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node7:
     connections:
@@ -76,13 +76,13 @@ nodes:
         index: 0
       node1:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node8:
     connections:
       node1:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node9:
     connections:
@@ -90,5 +90,5 @@ nodes:
         index: 0
       node5:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}

--- a/tests/functional/mesh/mesh-definitions/random-mesh-udp.yaml
+++ b/tests/functional/mesh/mesh-definitions/random-mesh-udp.yaml
@@ -2,13 +2,13 @@
 nodes:
   controller:
     connections: {}
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node1:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node10:
     connections:
@@ -16,13 +16,13 @@ nodes:
         index: 0
       node4:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node11:
     connections:
       node1:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node12:
     connections:
@@ -30,13 +30,13 @@ nodes:
         index: 0
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node2:
     connections:
       node1:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node3:
     connections:
@@ -44,7 +44,7 @@ nodes:
         index: 0
       node4:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node4:
     connections:
@@ -52,7 +52,7 @@ nodes:
         index: 0
       node2:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node5:
     connections:
@@ -60,7 +60,7 @@ nodes:
         index: 0
       node12:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node6:
     connections:
@@ -68,7 +68,7 @@ nodes:
         index: 0
       node3:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node7:
     connections:
@@ -76,13 +76,13 @@ nodes:
         index: 0
       node1:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node8:
     connections:
       node1:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node9:
     connections:
@@ -90,5 +90,5 @@ nodes:
         index: 0
       node5:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}

--- a/tests/functional/mesh/mesh-definitions/random-mesh-ws.yaml
+++ b/tests/functional/mesh/mesh-definitions/random-mesh-ws.yaml
@@ -2,13 +2,13 @@
 nodes:
   controller:
     connections: {}
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node1:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node10:
     connections:
@@ -16,13 +16,13 @@ nodes:
         index: 0
       node4:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node11:
     connections:
       node1:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node12:
     connections:
@@ -30,13 +30,13 @@ nodes:
         index: 0
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node2:
     connections:
       node1:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node3:
     connections:
@@ -44,7 +44,7 @@ nodes:
         index: 0
       node4:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node4:
     connections:
@@ -52,7 +52,7 @@ nodes:
         index: 0
       node2:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node5:
     connections:
@@ -60,7 +60,7 @@ nodes:
         index: 0
       node12:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node6:
     connections:
@@ -68,7 +68,7 @@ nodes:
         index: 0
       node3:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node7:
     connections:
@@ -76,13 +76,13 @@ nodes:
         index: 0
       node1:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node8:
     connections:
       node1:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node9:
     connections:
@@ -90,5 +90,5 @@ nodes:
         index: 0
       node5:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}

--- a/tests/functional/mesh/mesh-definitions/tree-mesh-tcp.yaml
+++ b/tests/functional/mesh/mesh-definitions/tree-mesh-tcp.yaml
@@ -2,77 +2,77 @@
 nodes:
   controller:
     connections: {}
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node1:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node10:
     connections:
       node9:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node11:
     connections:
       node9:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node12:
     connections:
       node9:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node2:
     connections:
       node1:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node3:
     connections:
       node1:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node4:
     connections:
       node1:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node5:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node6:
     connections:
       node5:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node7:
     connections:
       node5:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node8:
     connections:
       node5:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}
   node9:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - tcp-listener: {}

--- a/tests/functional/mesh/mesh-definitions/tree-mesh-udp.yaml
+++ b/tests/functional/mesh/mesh-definitions/tree-mesh-udp.yaml
@@ -2,77 +2,77 @@
 nodes:
   controller:
     connections: {}
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node1:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node10:
     connections:
       node9:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node11:
     connections:
       node9:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node12:
     connections:
       node9:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node2:
     connections:
       node1:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node3:
     connections:
       node1:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node4:
     connections:
       node1:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node5:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node6:
     connections:
       node5:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node7:
     connections:
       node5:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node8:
     connections:
       node5:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}
   node9:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - udp-listener: {}

--- a/tests/functional/mesh/mesh-definitions/tree-mesh-ws.yaml
+++ b/tests/functional/mesh/mesh-definitions/tree-mesh-ws.yaml
@@ -2,77 +2,77 @@
 nodes:
   controller:
     connections: {}
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node1:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node10:
     connections:
       node9:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node11:
     connections:
       node9:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node12:
     connections:
       node9:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node2:
     connections:
       node1:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node3:
     connections:
       node1:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node4:
     connections:
       node1:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node5:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node6:
     connections:
       node5:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node7:
     connections:
       node5:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node8:
     connections:
       node5:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}
   node9:
     connections:
       controller:
         index: 0
-    nodedef:
+    nodedefbase:
       - ws-listener: {}

--- a/tests/functional/mesh/mesh_test.go
+++ b/tests/functional/mesh/mesh_test.go
@@ -278,7 +278,7 @@ func TestCosts(t *testing.T) {
 	// if they exist
 	data.Nodes["node1"] = &mesh.YamlNode{
 		Connections: map[string]mesh.YamlConnection{},
-		Nodedef: []interface{}{
+		NodedefBase: []interface{}{
 			map[interface{}]interface{}{
 				"tcp-listener": map[interface{}]interface{}{
 					"cost": 4.5,
@@ -296,7 +296,7 @@ func TestCosts(t *testing.T) {
 				Index: 0,
 			},
 		},
-		Nodedef: []interface{}{},
+		NodedefBase: []interface{}{},
 	}
 	data.Nodes["node3"] = &mesh.YamlNode{
 		Connections: map[string]mesh.YamlConnection{
@@ -304,7 +304,7 @@ func TestCosts(t *testing.T) {
 				Index: 0,
 			},
 		},
-		Nodedef: []interface{}{},
+		NodedefBase: []interface{}{},
 	}
 	data.Nodes["node4"] = &mesh.YamlNode{
 		Connections: map[string]mesh.YamlConnection{
@@ -312,7 +312,7 @@ func TestCosts(t *testing.T) {
 				Index: 0,
 			},
 		},
-		Nodedef: []interface{}{},
+		NodedefBase: []interface{}{},
 	}
 	m, err := mesh.NewCLIMeshFromYaml(data, t.Name())
 	if err != nil {
@@ -355,7 +355,7 @@ func TestDuplicateNodes(t *testing.T) {
 	// if they exist
 	data.Nodes["node1"] = &mesh.YamlNode{
 		Connections: map[string]mesh.YamlConnection{},
-		Nodedef: []interface{}{
+		NodedefBase: []interface{}{
 			map[interface{}]interface{}{
 				"tcp-listener": map[interface{}]interface{}{},
 			},
@@ -367,7 +367,7 @@ func TestDuplicateNodes(t *testing.T) {
 				Index: 0,
 			},
 		},
-		Nodedef: []interface{}{
+		NodedefBase: []interface{}{
 			map[interface{}]interface{}{
 				"tcp-listener": map[interface{}]interface{}{},
 			},
@@ -379,7 +379,7 @@ func TestDuplicateNodes(t *testing.T) {
 				Index: 0,
 			},
 		},
-		Nodedef: []interface{}{
+		NodedefBase: []interface{}{
 			map[interface{}]interface{}{
 				"node": map[interface{}]interface{}{
 					"id": "node1",
@@ -421,7 +421,7 @@ func benchmarkLinearMeshStartup(totalNodes int, b *testing.B) {
 			}
 			data.Nodes[nodeID] = &mesh.YamlNode{
 				Connections: connections,
-				Nodedef: []interface{}{
+				NodedefBase: []interface{}{
 					map[interface{}]interface{}{
 						"tcp-listener": map[interface{}]interface{}{},
 					},

--- a/tests/functional/mesh/tls_test.go
+++ b/tests/functional/mesh/tls_test.go
@@ -44,7 +44,7 @@ func TestTCPSSLConnections(t *testing.T) {
 			// if they exist
 			data.Nodes["node1"] = &mesh.YamlNode{
 				Connections: map[string]mesh.YamlConnection{},
-				Nodedef: []interface{}{
+				NodedefBase: []interface{}{
 					map[interface{}]interface{}{
 						"tls-server": map[interface{}]interface{}{
 							"name":              "cert1",
@@ -68,7 +68,7 @@ func TestTCPSSLConnections(t *testing.T) {
 						TLS:   "client-cert2",
 					},
 				},
-				Nodedef: []interface{}{
+				NodedefBase: []interface{}{
 					map[interface{}]interface{}{
 						"tls-server": map[interface{}]interface{}{
 							"name": "server-cert2",
@@ -98,7 +98,7 @@ func TestTCPSSLConnections(t *testing.T) {
 						TLS:   "client-secure",
 					},
 				},
-				Nodedef: []interface{}{
+				NodedefBase: []interface{}{
 					map[interface{}]interface{}{
 						"tls-client": map[interface{}]interface{}{
 							"name":    "client-secure",
@@ -174,7 +174,7 @@ func TestTCPSSLClientAuthFailNoKey(t *testing.T) {
 			// if they exist
 			data.Nodes["node1"] = &mesh.YamlNode{
 				Connections: map[string]mesh.YamlConnection{},
-				Nodedef: []interface{}{
+				NodedefBase: []interface{}{
 					map[interface{}]interface{}{
 						"tls-server": map[interface{}]interface{}{
 							"name":              "cert1",
@@ -198,7 +198,7 @@ func TestTCPSSLClientAuthFailNoKey(t *testing.T) {
 						TLS:   "client-insecure",
 					},
 				},
-				Nodedef: []interface{}{
+				NodedefBase: []interface{}{
 					map[interface{}]interface{}{
 						"tls-client": map[interface{}]interface{}{
 							"name":               "client-insecure",
@@ -263,7 +263,7 @@ func TestTCPSSLClientAuthFailBadKey(t *testing.T) {
 			// if they exist
 			data.Nodes["node1"] = &mesh.YamlNode{
 				Connections: map[string]mesh.YamlConnection{},
-				Nodedef: []interface{}{
+				NodedefBase: []interface{}{
 					map[interface{}]interface{}{
 						"tls-server": map[interface{}]interface{}{
 							"name":              "cert1",
@@ -287,7 +287,7 @@ func TestTCPSSLClientAuthFailBadKey(t *testing.T) {
 						TLS:   "client-insecure",
 					},
 				},
-				Nodedef: []interface{}{
+				NodedefBase: []interface{}{
 					map[interface{}]interface{}{
 						"tls-client": map[interface{}]interface{}{
 							"name":               "client-insecure",
@@ -347,7 +347,7 @@ func TestTCPSSLServerAuthFailNoKey(t *testing.T) {
 			// if they exist
 			data.Nodes["node1"] = &mesh.YamlNode{
 				Connections: map[string]mesh.YamlConnection{},
-				Nodedef: []interface{}{
+				NodedefBase: []interface{}{
 					map[interface{}]interface{}{
 						listener: map[interface{}]interface{}{},
 					},
@@ -360,7 +360,7 @@ func TestTCPSSLServerAuthFailNoKey(t *testing.T) {
 						TLS:   "client-secure",
 					},
 				},
-				Nodedef: []interface{}{
+				NodedefBase: []interface{}{
 					map[interface{}]interface{}{
 						"tls-client": map[interface{}]interface{}{
 							"name":    "client-secure",
@@ -425,7 +425,7 @@ func TestTCPSSLServerAuthFailBadKey(t *testing.T) {
 			// if they exist
 			data.Nodes["node1"] = &mesh.YamlNode{
 				Connections: map[string]mesh.YamlConnection{},
-				Nodedef: []interface{}{
+				NodedefBase: []interface{}{
 					map[interface{}]interface{}{
 						"tls-server": map[interface{}]interface{}{
 							"name": "cert1",
@@ -447,7 +447,7 @@ func TestTCPSSLServerAuthFailBadKey(t *testing.T) {
 						TLS:   "client-secure",
 					},
 				},
-				Nodedef: []interface{}{
+				NodedefBase: []interface{}{
 					map[interface{}]interface{}{
 						"tls-client": map[interface{}]interface{}{
 							"name":    "client-secure",

--- a/tests/functional/mesh/work_test.go
+++ b/tests/functional/mesh/work_test.go
@@ -49,6 +49,13 @@ func TestWork(t *testing.T) {
 			"params":   "-c \"for i in {1..5}; do echo $i;done\"",
 		},
 	}
+	echoSleepLong50 := map[interface{}]interface{}{
+		"work-command": map[interface{}]interface{}{
+			"workType": "echosleeplong50",
+			"command":  "bash",
+			"params":   "-c \"for i in {1..50}; do echo $i; sleep 4;done\"",
+		},
+	}
 	kubeEchoSleepLong := map[interface{}]interface{}{
 		"work-kubernetes": map[interface{}]interface{}{
 			"workType":   "echosleeplong",
@@ -57,6 +64,16 @@ func TestWork(t *testing.T) {
 			"kubeconfig": filepath.Join(home, ".kube/config"),
 			"image":      "alpine",
 			"command":    "sh -c \"for i in `seq 1 5`; do echo $i; sleep 3;done\"",
+		},
+	}
+	kubeEchoSleepLong50 := map[interface{}]interface{}{
+		"work-kubernetes": map[interface{}]interface{}{
+			"workType":   "echosleeplong50",
+			"authmethod": "kubeconfig",
+			"namespace":  "default",
+			"kubeconfig": filepath.Join(home, ".kube/config"),
+			"image":      "alpine",
+			"command":    "sh -c \"for i in `seq 1 50`; do echo $i; sleep 4;done\"",
 		},
 	}
 	kubeEchoSleepShort := map[interface{}]interface{}{
@@ -70,17 +87,19 @@ func TestWork(t *testing.T) {
 		},
 	}
 	testTable := []struct {
-		testGroup    string
-		shortCommand map[interface{}]interface{}
-		longCommand  map[interface{}]interface{}
+		testGroup     string
+		shortCommand  map[interface{}]interface{}
+		longCommand   map[interface{}]interface{}
+		longCommand50 map[interface{}]interface{}
 	}{
-		{"normal_worker", echoSleepShort, echoSleepLong},
-		{"kube_worker", kubeEchoSleepShort, kubeEchoSleepLong},
+		{"normal_worker", echoSleepShort, echoSleepLong, echoSleepLong50},
+		{"kube_worker", kubeEchoSleepShort, kubeEchoSleepLong, kubeEchoSleepLong50},
 	}
 	for _, subtest := range testTable {
 		testGroup := subtest.testGroup
 		shortCommand := subtest.shortCommand
 		longCommand := subtest.longCommand
+		longCommand50 := subtest.longCommand50
 		// Setup our mesh yaml data
 		workSetup := func(testName string) (map[string]*receptorcontrol.ReceptorControl, *mesh.CLIMesh, []byte) {
 			data := mesh.YamlData{}
@@ -106,7 +125,7 @@ func TestWork(t *testing.T) {
 			// Generate a mesh with 3 nodes
 			data.Nodes["node2"] = &mesh.YamlNode{
 				Connections: map[string]mesh.YamlConnection{},
-				Nodedef: []interface{}{
+				NodedefBase: []interface{}{
 					map[interface{}]interface{}{
 						"tcp-listener": map[interface{}]interface{}{
 							"cost": 1.0,
@@ -140,7 +159,7 @@ func TestWork(t *testing.T) {
 						Index: 0,
 					},
 				},
-				Nodedef: []interface{}{
+				NodedefBase: []interface{}{
 					map[interface{}]interface{}{
 						"tls-client": map[interface{}]interface{}{
 							"name":               "tlsclient",
@@ -167,9 +186,10 @@ func TestWork(t *testing.T) {
 						Index: 0,
 					},
 				},
-				Nodedef: []interface{}{
+				NodedefBase: []interface{}{
 					longCommand,
 					shortCommand,
+					longCommand50,
 				},
 			}
 
@@ -587,6 +607,142 @@ func TestWork(t *testing.T) {
 				t.Fatal(err)
 			}
 		})
+		t.Run(testGroup+"/reload backends while streaming work results", func(t *testing.T) {
+			t.Parallel()
+			if strings.Contains(t.Name(), "kube") {
+				checkSkipKube(t)
+			}
+			controllers, m, _ := workSetup(t.Name())
+			defer tearDown(controllers, m)
+
+			// submit work from node 2 to node 3
+			unitID, err := controllers["node2"].WorkSubmit("node3", "echosleeplong50")
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			// wait for 10 seconds, and check if the work is in pending state
+			ctx, _ := context.WithTimeout(context.Background(), 10*time.Second)
+			err = controllers["node2"].AssertWorkPending(ctx, unitID)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			// declare a new mesh with no connection between nodes
+			modifiedData := mesh.YamlData{}
+			modifiedData.Nodes = make(map[string]*mesh.YamlNode)
+			modifiedData.Nodes["node2"] = &mesh.YamlNode{
+				Connections: map[string]mesh.YamlConnection{},
+			}
+			modifiedData.Nodes["node1"] = &mesh.YamlNode{
+				Connections: map[string]mesh.YamlConnection{},
+			}
+			modifiedData.Nodes["node3"] = &mesh.YamlNode{
+				Connections: map[string]mesh.YamlConnection{},
+			}
+
+			// modify the existing mesh
+			err = mesh.ModifyCLIMeshFromYaml(modifiedData, *m)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			// reload the entire mesh
+			for k := range controllers {
+				err = controllers[k].Reload()
+				if err != nil {
+					t.Fatal(err)
+				}
+			}
+
+			// ping should not be successful in a broken mesh
+			_, err = controllers["node1"].Ping("node3")
+			if err == nil {
+				t.Fatal("ping should not be successful")
+			}
+
+			// read the work status
+			workStatus, err := controllers["node2"].GetWorkStatus(unitID)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			// modify the mesh to have connection again
+			withConnectionData := mesh.YamlData{}
+			withConnectionData.Nodes = make(map[string]*mesh.YamlNode)
+			withConnectionData.Nodes["node2"] = &mesh.YamlNode{
+				Connections: map[string]mesh.YamlConnection{},
+			}
+			withConnectionData.Nodes["node1"] = &mesh.YamlNode{
+				Connections: map[string]mesh.YamlConnection{
+					"node2": {
+						Index: 0,
+					},
+				},
+			}
+			withConnectionData.Nodes["node3"] = &mesh.YamlNode{
+				Connections: map[string]mesh.YamlConnection{
+					"node2": {
+						Index: 0,
+					},
+				},
+			}
+
+			err = mesh.ModifyCLIMeshFromYaml(withConnectionData, *m)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			// reload the entire mesh
+			for k := range controllers {
+				err = controllers[k].Reload()
+				if err != nil {
+					t.Fatal(err)
+				}
+			}
+
+			// wait for mesh to become ready again
+			ctx, _ = context.WithTimeout(context.Background(), 60*time.Second)
+			err = m.WaitForReady(ctx)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			// ping should be successful in the mesh with connections
+			_, err = controllers["node1"].Ping("node3")
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			ctx, _ = context.WithTimeout(context.Background(), 15*time.Second)
+			err = controllers["node2"].AssertWorkSizeIncreasing(ctx, unitID, workStatus.StdoutSize)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			// it takes some time for the streaming to start again
+			time.Sleep(15 * time.Second)
+
+			newWorkStatus, err := controllers["node2"].GetWorkStatus(unitID)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			if newWorkStatus.StdoutSize <= workStatus.StdoutSize {
+				t.Fatal("Work size is not increasing")
+			}
+
+			// cancel the work so that it doesnt run after the test ends
+			_, err = controllers["node2"].WorkCancel(unitID)
+			if err != nil {
+				t.Fatal(err)
+			}
+			ctx, _ = context.WithTimeout(context.Background(), 15*time.Second)
+			err = controllers["node2"].AssertWorkCancelled(ctx, unitID)
+			if err != nil {
+				t.Fatal(err)
+			}
+		})
 	}
 }
 
@@ -602,7 +758,7 @@ func TestRuntimeParams(t *testing.T) {
 	data.Nodes = make(map[string]*mesh.YamlNode)
 	data.Nodes["node0"] = &mesh.YamlNode{
 		Connections: map[string]mesh.YamlConnection{},
-		Nodedef: []interface{}{
+		NodedefBase: []interface{}{
 			map[interface{}]interface{}{
 				"tcp-listener": map[interface{}]interface{}{},
 			},
@@ -664,7 +820,7 @@ func TestKubeRuntimeParams(t *testing.T) {
 	data.Nodes = make(map[string]*mesh.YamlNode)
 	data.Nodes["node0"] = &mesh.YamlNode{
 		Connections: map[string]mesh.YamlConnection{},
-		Nodedef: []interface{}{
+		NodedefBase: []interface{}{
 			map[interface{}]interface{}{
 				"tcp-listener": map[interface{}]interface{}{},
 			},
@@ -723,7 +879,7 @@ func TestRuntimeParamsNotAllowed(t *testing.T) {
 	data.Nodes = make(map[string]*mesh.YamlNode)
 	data.Nodes["node0"] = &mesh.YamlNode{
 		Connections: map[string]mesh.YamlConnection{},
-		Nodedef: []interface{}{
+		NodedefBase: []interface{}{
 			map[interface{}]interface{}{
 				"tcp-listener": map[interface{}]interface{}{},
 			},
@@ -776,7 +932,7 @@ func TestKubeContainerFailure(t *testing.T) {
 	data.Nodes = make(map[string]*mesh.YamlNode)
 	data.Nodes["node0"] = &mesh.YamlNode{
 		Connections: map[string]mesh.YamlConnection{},
-		Nodedef: []interface{}{
+		NodedefBase: []interface{}{
 			map[interface{}]interface{}{
 				"tcp-listener": map[interface{}]interface{}{},
 			},

